File: src/main/java/dev/slne/discord/ticket/Ticket.java
package dev.slne.discord.ticket;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import dev.slne.discord.DiscordBot;
import dev.slne.discord.config.discord.GuildConfig;
import dev.slne.discord.config.role.RoleConfig;
import dev.slne.discord.discord.interaction.command.commands.ticket.TicketCloseCommand;
import dev.slne.discord.message.Messages;
import dev.slne.discord.ticket.member.TicketMember;
import dev.slne.discord.ticket.message.TicketMessage;
import java.time.ZonedDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import lombok.AccessLevel;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.ToString;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.PermissionOverride;
import net.dv8tion.jda.api.entities.Role;
import net.dv8tion.jda.api.entities.User;
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel;
import net.dv8tion.jda.api.requests.RestAction;
import net.dv8tion.jda.api.requests.restaction.PermissionOverrideAction;
import org.jetbrains.annotations.Blocking;

/**
 * The type Ticket.
 */
@Getter
@ToString
@EqualsAndHashCode
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Ticket {

  private final List<TicketMember> removedMembers = new ArrayList<>();

  @JsonProperty("id")
  @Setter
  private long id = -1;

  @JsonProperty("ticket_id")
  @Setter
  private UUID ticketId;

  @JsonProperty("opened_at")
  @Setter
  private ZonedDateTime openedAt;

  @JsonProperty("guild_id")
  private String guildId;

  @JsonProperty("channel_id")
  @Setter
  private String channelId;

  @JsonProperty("type")
  private String ticketTypeString;

  @JsonProperty("author_id")
  private String ticketAuthorId;

  @JsonProperty("author_name")
  private String ticketAuthorName;

  @JsonProperty("author_avatar_url")
  private String ticketAuthorAvatarUrl;

  @JsonProperty("closed_by_id")
  @Setter
  private String closedById;

  @JsonProperty("closed_reason")
  @Setter
  private String closedReason;

  @JsonProperty("closed_by_avatar_url")
  @Setter
  private String closedByAvatarUrl;

  @JsonProperty("closed_by_name")
  @Setter
  private String closedByName;

  @JsonProperty("closed_at")
  @Setter
  private ZonedDateTime closedAt;

  @JsonProperty("messages")
  private List<TicketMessage> messages = new ArrayList<>();

  @JsonProperty("members")
  private List<TicketMember> members = new ArrayList<>();

  @JsonProperty("created_at")
  @Setter
  private ZonedDateTime createdAt;

  /**
   * Constructor for a ticket
   *
   * @param guild        The guild the ticket is created in
   * @param ticketAuthor The author of the ticket
   * @param ticketType   The type of the ticket
   */
  public Ticket(Guild guild, User ticketAuthor, TicketType ticketType) {
    this.openedAt = ZonedDateTime.now();

    if (guild != null) {
      this.guildId = guild.getId();
    }

    if (ticketType != null) {
      this.ticketTypeString = ticketType.name();
    }

    if (ticketAuthor != null) {
      this.ticketAuthorName = ticketAuthor.getName();
      this.ticketAuthorId = ticketAuthor.getId();
      this.ticketAuthorAvatarUrl = ticketAuthor.getAvatarUrl();
    }
  }

  /**
   * Adds a raw ticket member to the ticket
   *
   * @param ticketMember The ticket member
   */
  public void addRawTicketMember(TicketMember ticketMember) {
    members.add(ticketMember);
  }

  /**
   * Removes a raw ticket member from the ticket
   *
   * @param ticketMember The ticket member
   */
  public void removeRawTicketMember(TicketMember ticketMember) {
    members.remove(ticketMember);
    removedMembers.add(ticketMember);
  }

  /**
   * Adds a raw ticket message to the ticket
   *
   * @param ticketMessage The ticket message
   */
  public void addRawTicketMessage(TicketMessage ticketMessage) {
    messages.add(ticketMessage);
  }

  /**
   * Check if the member exists
   *
   * @param user The user to check
   * @return If the member exists
   */
  @JsonIgnore
  public boolean memberExists(User user) {
    return members.stream()
        .anyMatch(member -> member.getMemberId().equals(user.getId()) && !member.isRemoved());
  }


  /**
   * Adds a role to the ticket channel
   *
   * @param role The role
   * @return The permission override
   */
  public CompletableFuture<PermissionOverride> addTicketRole(Role role) { // TODO: 25.08.2024 15:24 - needed?
    Optional<TextChannel> channel = getChannel();

    if (channel.isEmpty()) {
      return CompletableFuture.completedFuture(null);
    }

    GuildConfig guildConfig = GuildConfig.getConfig(getGuildId());

    if (guildConfig == null) {
      return CompletableFuture.completedFuture(null);
    }

    RoleConfig roleConfig = RoleConfig.getConfig(getGuildId(), role.getName());

    PermissionOverrideAction permissionOverrideAction = channel.get().upsertPermissionOverride(role);
    permissionOverrideAction = permissionOverrideAction.resetAllow();
    permissionOverrideAction = permissionOverrideAction.resetDeny();
    permissionOverrideAction = permissionOverrideAction.setAllowed(
        roleConfig.getDiscordDeniedPermissionsAsJDA());

    return permissionOverrideAction.submit();
  }

  /**
   * Returns the ticket message by the message
   *
   * @param message The message
   * @return The ticket message
   */
  @SuppressWarnings("unused")
  @JsonIgnore
  public Optional<TicketMessage> getTicketMessage(Message message) {
    return messages.stream()
        .filter(ticketMessage -> ticketMessage.getMessageId().equals(message.getId()))
        .findFirst();
  }

  /**
   * Returns the ticket message by the message id
   *
   * @param messageId The message id
   * @return The ticket message
   */
  @JsonIgnore
  public Optional<TicketMessage> getTicketMessage(String messageId) {
    return messages.stream().filter(ticketMessage -> ticketMessage.getMessageId().equals(messageId))
        .findFirst();
  }

  /**
   * Returns the ticket member by the member
   *
   * @param user The member
   * @return The ticket member
   */
  @SuppressWarnings("unused")
  @JsonIgnore
  public Optional<TicketMember> getTicketMember(User user) {
    return members.stream().filter(ticketMember -> ticketMember.getMemberId().equals(user.getId()))
        .findFirst();
  }

  /**
   * Returns the active ticket member by the member
   *
   * @param user The member
   * @return The active ticket member
   */
  @JsonIgnore
  public Optional<TicketMember> getActiveTicketMember(User user) {
    return members.stream()
        .filter(ticketMember -> ticketMember.getMemberId().equals(user.getId()) &&
            !ticketMember.isRemoved())
        .findFirst();
  }

  /**
   * Returns the ticket member by the member id
   *
   * @param userId The member id
   * @return The ticket member
   */
  @SuppressWarnings("unused")
  @JsonIgnore
  public Optional<TicketMember> getTicketMember(String userId) {
    return members.stream()
        .filter(ticketMember -> ticketMember.getMemberId().equals(userId))
        .findFirst();
  }

  /**
   * Gets ticket author.
   *
   * @return the ticketAuthor
   */
  @JsonIgnore
  public Optional<RestAction<User>> getTicketAuthor() {
    if (ticketAuthorId == null) {
      return Optional.empty();
    }

    return Optional.of(DiscordBot.getInstance().getJda().retrieveUserById(ticketAuthorId));
  }

  @JsonIgnore
  @Blocking
  public Optional<User> getTicketAuthorNow() {
    return getTicketAuthor().map(RestAction::complete);
  }

  /**
   * Get the type of the ticket
   *
   * @return The type of the ticket
   */
  @JsonIgnore
  public TicketType getTicketType() {
    return TicketType.valueOf(ticketTypeString);
  }

  /**
   * Get the guild the ticket is created in
   *
   * @return The guild the ticket is created in
   */
  @JsonIgnore
  public Optional<Guild> getGuild() {
    if (guildId == null) {
      return Optional.empty();
    }

    return Optional.ofNullable(DiscordBot.getInstance().getJda().getGuildById(guildId));
  }

  /**
   * Get the channel the ticket is created in
   *
   * @return The channel the ticket is created in
   */
  @JsonIgnore
  public Optional<TextChannel> getChannel() {
    if (channelId == null) {
      return Optional.empty();
    }

    return Optional.ofNullable(DiscordBot.getInstance().getJda().getTextChannelById(channelId));
  }

  /**
   * Gets closed by.
   *
   * @return the closedBy
   */
  @JsonIgnore
  public Optional<RestAction<User>> getClosedBy() {
    if (closedById == null) {
      return Optional.empty();
    }

    return Optional.of(DiscordBot.getInstance().getJda().retrieveUserById(closedById));
  }

  @JsonIgnore
  public Optional<User> getClosedByNow() {
    return getClosedBy().map(RestAction::complete);
  }

  public void updateFrom(Ticket other) {
    setOpenedAt(other.getOpenedAt());
    setTicketId(other.getTicketId());
    setId(other.getId());
    setCreatedAt(other.getCreatedAt());
    setClosedAt(other.getClosedAt());
  }

  @JsonIgnore
  public boolean isPersisted() {
    return id != -1;
  }

  @JsonIgnore
  public boolean hasTicketId() {
    return ticketId != null;
  }

  @JsonIgnore
  public boolean hasGuild() {
    return guildId != null;
  }

  @JsonIgnore
  public String getCloseReasonOrDefault() {
    return closedReason == null ? Messages.DEFAULT_TICKET_CLOSED_REASON : closedReason;
  }
}

File: src/main/kotlin/dev/slne/discord/ticket/Ticket.kt
package dev.slne.discord.ticket

import dev.slne.discord.getBean
import dev.slne.discord.message.Messages
import dev.slne.discord.ticket.message.TicketMessage
import dev.slne.discord.util.freeze
import dev.slne.discord.util.toObjectList
import jakarta.persistence.*
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.Message
import net.dv8tion.jda.api.entities.User
import org.hibernate.annotations.JdbcTypeCode
import org.hibernate.annotations.TimeZoneStorage
import org.hibernate.annotations.TimeZoneStorageType
import org.hibernate.proxy.HibernateProxy
import org.hibernate.type.SqlTypes
import java.time.ZonedDateTime
import java.util.*

@Entity
@Table(name = "ticket_tickets")
data class Ticket(
    @Id
    @JdbcTypeCode(SqlTypes.BIGINT)
    @Column(name = "id", nullable = false)
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null,

    @JdbcTypeCode(SqlTypes.CHAR)
    @Column(name = "ticket_id", nullable = false, unique = true, length = 36)
    val ticketId: UUID,

    @JdbcTypeCode(SqlTypes.CHAR)
    @Enumerated(EnumType.STRING)
    @Column(name = "ticket_type")
    val ticketType: TicketType,

    @Transient
    val ticketAuthor: User,

    @TimeZoneStorage(TimeZoneStorageType.NORMALIZE_UTC)
    @Column(name = "opened_at", nullable = false)
    val openedAt: ZonedDateTime,

    @Transient
    val internalGuild: Guild
) {

    constructor(
        guild: Guild,
        author: User,
        ticketType: TicketType,
    ) : this(
        ticketId = UUID.randomUUID(),
        ticketType = ticketType,
        ticketAuthor = author,
        openedAt = ZonedDateTime.now(),
        internalGuild = guild
    )

    @JdbcTypeCode(SqlTypes.CHAR)
    @Column(name = "guild_id", length = 20)
    val guildId: String = internalGuild.id

    @JdbcTypeCode(SqlTypes.CHAR)
    @Column(name = "thread_id", length = 20)
    var threadId: String? = null

    @JdbcTypeCode(SqlTypes.CHAR)
    @Column(name = "ticket_author_id", length = 20, nullable = false)
    val ticketAuthorId = ticketAuthor.id

    @JdbcTypeCode(SqlTypes.CHAR)
    @Column(name = "ticket_author_name", length = 64, nullable = false)
    val ticketAuthorName = ticketAuthor.name

    @JdbcTypeCode(SqlTypes.CHAR)
    @Column(name = "ticket_author_avatar_url")
    val ticketAuthorAvatarUrl = ticketAuthor.avatarUrl

    @JdbcTypeCode(SqlTypes.CHAR)
    @Column(name = "closed_by_id", length = 20)
    var closedById: String? = null

    @Lob
    @Column(name = "closed_reason")
    var closedReason: String? = null

    @JdbcTypeCode(SqlTypes.CHAR)
    @Column(name = "closed_by_avatar_url")
    var closedByAvatarUrl: String? = null

    @JdbcTypeCode(SqlTypes.CHAR)
    @Column(name = "closed_by_name", length = 32)
    var closedByName: String? = null

    @TimeZoneStorage(TimeZoneStorageType.NORMALIZE_UTC)
    @Column(name = "closed_at")
    var closedAt: ZonedDateTime? = null

    @Transient
    var isClosing = false

    @OneToMany(
        mappedBy = "ticket",
        cascade = [CascadeType.ALL],
        fetch = FetchType.EAGER,
    )
    private var _messages = mutableListOf<TicketMessage>()
    val messages get() = _messages.toObjectList().freeze()

    val thread get() = threadId?.let { getBean<JDA>().getThreadChannelById(it) }
    val closedBy get() = closedById?.let { getBean<JDA>().retrieveUserById(it) }
    val closeReasonOrDefault: String get() = closedReason ?: Messages.DEFAULT_TICKET_CLOSED_REASON
    val author get() = getBean<JDA>().retrieveUserById(ticketAuthorId)
    val guild get() = getBean<JDA>().getGuildById(guildId)

    val isClosed
        get() = closedAt != null

    fun addMessage(message: TicketMessage) {
        _messages.add(message)
        message.ticket = this
    }

    fun getTicketMessage(message: Message) = getTicketMessage(message.id)
    fun getTicketMessage(messageId: String) = _messages.firstOrNull { it.messageId == messageId }

    fun close(
        closedBy: User,
        closedReason: String,
        closedAt: ZonedDateTime = ZonedDateTime.now()
    ) {
        this.closedById = closedBy.id
        this.closedByName = closedBy.name
        this.closedByAvatarUrl = closedBy.avatarUrl
        this.closedReason = closedReason
        this.closedAt = closedAt
    }

    fun reopen() {
        this.closedById = null
        this.closedByName = null
        this.closedByAvatarUrl = null
        this.closedReason = null
        this.closedAt = null
    }

    final override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null) return false
        val oEffectiveClass =
            if (other is HibernateProxy) other.hibernateLazyInitializer.persistentClass else other.javaClass
        val thisEffectiveClass =
            if (this is HibernateProxy) this.hibernateLazyInitializer.persistentClass else this.javaClass
        if (thisEffectiveClass != oEffectiveClass) return false
        other as Ticket

        return id != null && id == other.id
    }

    final override fun hashCode(): Int =
        if (this is HibernateProxy) this.hibernateLazyInitializer.persistentClass.hashCode() else javaClass.hashCode()

    override fun toString(): String {
        return "Ticket(_messages=$_messages, isClosing=$isClosing, closedAt=$closedAt, closedByName=$closedByName, closedByAvatarUrl=$closedByAvatarUrl, closedReason=$closedReason, closedById=$closedById, ticketAuthorAvatarUrl=$ticketAuthorAvatarUrl, ticketAuthorName='$ticketAuthorName', ticketAuthorId='$ticketAuthorId', threadId=$threadId, guildId='$guildId', internalGuild=$internalGuild, openedAt=$openedAt, ticketAuthor=$ticketAuthor, ticketType=$ticketType, ticketId=$ticketId, id=$id)"
    }

}